### ISSUE TITLE
博主 求解释

### ISSUE THREAD

**Issue by khazic** [chenjunyi1999/NER-Medical#1]
你好 我想问一下 你这个是bert-base-chinese 那我现在把process_data该成英文的bio，我把代码中的bert_chinese换成bert_base_uncased 为什么运行不了呢 他可以改成英文的模型吗

**Comment by chenjunyi1999** [comment]
您好，理论上改成英文模型是没有问题的。这边有两个想法
1. 英文BIO比中文更加复杂，需要先转化成token之后再进行BIO处理。不知道这个您处理的是否正确。这个比赛[feedback-prize-2021](https://www.kaggle.com/competitions/feedback-prize-2021/)的讨论区上有很多不错的英文处理代码可供参考。

2. 除此之外“运行不了”有没有具体报错信息？如果有的话可以按照报错信息进行debug，如果没有报错只是训练后模型推理不正确，则可以检查梯度，调整学习率之类的。
希望对你有帮助

**Comment by khazic** [comment]
你好 我现在运行成功了 我正在尝试下载一些英文的训练集来做测试

**Comment by khazic** [comment]
但是我发现一个问题 就是他很容易把一个单词拆分开来 比如in Beverly Hills, California  他会分成beverly h    cali。。。。

**Comment by khazic** [comment]
这个我知道是英文bert本身最开始设计上的缺陷 有没有什么办法改成英文版的wwm the whole word masking呢 因为这样的情况还比较多

**Comment by chenjunyi1999** [comment]
subword tokenization会尽量将不常见的词拆分为更常见的词根组合（例如“Transformer”拆分为'transform'和 '##er'），如果你不想这样，可以将tokenzier的`is_split_into_words` 设置为`True`。您可以看一下这个下面的这个[文档](https://huggingface.co/docs/transformers/v4.28.1/en/main_classes/tokenizer#tokenizer)，希望对您有帮助。 😊

**Comment by khazic** [comment]
我现在有遇到一个问题 我拿了几千组的数据去训练 50个epoch 训练的是英文版的城市 组织 国家 
valid-->Epoch: 50, Val Loss:74.4708, Val Acc:87.422%
              precision    recall  f1-score   support

       B-INS     0.7982    0.7982    0.7982      1947
       B-COU     0.9142    0.9521    0.9328      1253
       B-CIT     0.8485    0.8276    0.8379       812
       I-INS     0.7690    0.8293    0.7980      4511
       I-COU     0.9048    0.9135    0.9091       104
       I-CIT     0.9008    0.9083    0.9046       120

   micro avg     0.8061    0.8419    0.8236      8747
   macro avg     0.8559    0.8715    0.8634      8747
weighted avg     0.8071    0.8419    0.8238      8747



但是为什么测试的时候居然连usc都没能学到呢

**Comment by chenjunyi1999** [comment]
这个可能需要你自己检查一下，可以先检查模型参数是否成功的被保存到ckpt，再根据训练结果和推理结果看看是否学习率需要调整。

**Comment by khazic** [comment]
好的 按照你刚才说的方法是将is_split_into_words加入到代码中

但是他报错TypeError: convert_tokens_to_ids() got an unexpected keyword argument 'is_split_into_words'

我是加在：
    def __getitem__(self, idx):
        sentence, label = self.sentences[idx], self.labels[idx]
        sentence_ids = self.tokenizer.convert_tokens_to_ids(sentence,is_split_into_words=True)
        label_ids = [tag2idx[l] for l in label]
        seqlen = len(label_ids)
        return sentence_ids, label_ids, seqlen
不是加载这里吗还是说加在self.tokenizer = BertTokenizer.from_pretrained(BASE_MODEL)呢 感谢解答

**Comment by khazic** [comment]
不好意思 我写错位了 解决了！感谢

**Comment by chenjunyi1999** [comment]
嗯嗯，好的。那我关闭这个issue了。

**Comment by khazic** [comment]
稍等 等我出一下结果可以吗 我短暂的跑十个epoch

**Comment by chenjunyi1999** [comment]
好👌

**Comment by khazic** [comment]
博主 我训练了50个epoch之后 f1=89 loss==0.5左右
(khazic) [ec2-user@ip-172-31-39-57 NER-Medical-main]$ sudo python3 main.py --mode='infer' --ckpt_name="best" --txt="usa"
Start infer
Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertModel: ['cls.predictions.transform.dense.bias', 'cls.predictions.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.decoder.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.transform.LayerNorm.weight']
- This IS expected if you are initializing BertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Loading the pre-trained checkpoint...
The result are shown below:
[['国家', ['u'], 1, 2]]
entity_name:u, entity_type:国家, start_pos:1, end_pos:2
(khazic) [ec2-user@ip-172-31-39-57 NER-Medical-main]$ 


然后那个split的代码我也加了 但是预测的结果还是很不好

**Comment by chenjunyi1999** [comment]
我之前做的时候可能没有这个warning，但我认为这并不是一个报错，很多问题可以在stackoverflow上找到答案[链接](https://stackoverflow.com/questions/67546911/python-bert-error-some-weights-of-the-model-checkpoint-at-were-not-used-when)
有非常多的原因可能会导致预测结果比较差，这个需要你根据你的代码自行检查。
